prov/efa/src/efa_mr.c

```diff
@@ -511,76 +521,67 @@ struct ibv_mr *efa_mr_reg_ibv_dmabuf_mr(struct ibv_pd *pd, uint64_t offset,
+
+	dmabuf_support = &(efa_mr->domain->hmem_info[mr_attr->iface].dmabuf_support_status);
+
+	if (flags & FI_MR_DMABUF) {
+		goto dmabuf;
+	}
+
+	if (*dmabuf_support == EFA_DMABUF_NOT_SUPPORTED) {
+		goto no_dmabuf;
+	}
+dmabuf:
+	if (flags & FI_MR_DMABUF) {
+		dmabuf_fd = mr_attr->dmabuf->fd;
+		offset = mr_attr->dmabuf->offset;
+		iova = (uintptr_t) mr_attr->dmabuf->base_addr +
+		       mr_attr->dmabuf->offset;
+		len = mr_attr->dmabuf->len;
+	} else {
+		iova = (uint64_t) mr_attr->mr_iov->iov_base,
+		len = mr_attr->mr_iov->iov_len;
+		ret = ofi_hmem_get_dmabuf_fd(mr_attr->iface, (void *)iova, len,
+		                             &dmabuf_fd, &offset);
 
-	/*
-	 * TODO: remove the synapseai and neuron blocks by onboarding the
-	 * ofi_hmem_get_dmabuf_fd API.
-	 */
-#if HAVE_SYNAPSEAI
-	if (efa_mr_is_synapseai(efa_mr)) {
-		int dmabuf_fd;
-		uint64_t offset;
-		int ret;
-
-		ret = synapseai_get_dmabuf_fd(mr_attr->mr_iov->iov_base,
-						(uint64_t) mr_attr->mr_iov->iov_len,
-						&dmabuf_fd, &offset);
 		if (ret != FI_SUCCESS) {
-			EFA_WARN(FI_LOG_MR, "Unable to get dmabuf fd for Gaudi device buffer \n");
-			return NULL;
+			EFA_WARN(FI_LOG_MR,
+				 "Unable to get dmabuf fd for device buffer. "
+				 "errno: %d, err_msg: %s\n",
+				 ret, fi_strerror(-ret));
+			goto fallback;
 		}
-		return efa_mr_reg_ibv_dmabuf_mr(efa_mr->domain->ibv_pd, offset,
-					mr_attr->mr_iov->iov_len,
-					(uint64_t)mr_attr->mr_iov->iov_base,
-					dmabuf_fd, access);
 	}
-#endif
 
-#if HAVE_NEURON
-	if (efa_mr_is_neuron(efa_mr)) {
-		int dmabuf_fd;
-		uint64_t offset;
-		int ret;
-
-		ret = neuron_get_dmabuf_fd(
-				mr_attr->mr_iov->iov_base,
-				mr_attr->mr_iov->iov_len,
-				&dmabuf_fd,
-				&offset);
-
-		if (ret == FI_SUCCESS) {
-			/* Success => invoke ibv_reg_dmabuf_mr */
-			return efa_mr_reg_ibv_dmabuf_mr(
-					efa_mr->domain->ibv_pd, 0,
-					mr_attr->mr_iov->iov_len,
-					(uint64_t)mr_attr->mr_iov->iov_base,
-					dmabuf_fd, access);
-		} else if (ret == -FI_ENOPROTOOPT) {
-			/* Protocol not availabe => fallback */
-			EFA_INFO(FI_LOG_MR,
-				"Unable to get dmabuf fd for Neuron device buffer, "
-				"Fall back to ibv_reg_mr\n");
-			return ibv_reg_mr(
-				efa_mr->domain->ibv_pd,
-				(void *)mr_attr->mr_iov->iov_base,
-				mr_attr->mr_iov->iov_len, access);
-		}
-		return NULL;
+	EFA_INFO(FI_LOG_MR,
+		 "Registering dmabuf mr with fd: %d, offset: %lu, len: %zu\n",
+		 dmabuf_fd, offset, len);
+
+	mr = efa_mr_reg_ibv_dmabuf_mr(efa_mr->domain->ibv_pd, offset, len, iova,
+				      dmabuf_fd, access);
+
+fallback:
+	if (*dmabuf_support == EFA_DMABUF_UNINITIALIZED) {
+		*dmabuf_support = mr == NULL ? EFA_DMABUF_NOT_SUPPORTED :
+		                               EFA_DMABUF_SUPPORTED;
+	}
+
+	if (flags & FI_MR_DMABUF || *dmabuf_support == EFA_DMABUF_SUPPORTED) {
+		goto out;
 	}
-#endif
 
-	return ibv_reg_mr(efa_mr->domain->ibv_pd,
-			(void *)mr_attr->mr_iov->iov_base,
+	EFA_INFO(FI_LOG_MR, "Disable dmabuf support for HMEM interface.\n");
+no_dmabuf:
```

Comment:
These goto are confusing, I think just doing them explicitly in the if blocks are better. See comments above.

Reply:
Yeah let's decide on the fallback logic first then I can simplify the logic here.

Reply:
Aggregate response to the comments around the fallback logic.

I have experimented with a couple different implementations, trying to reduce `goto` statements and improve the readability. But after all I figured this logic is the most intuitive(pseudo code):
```
efa_mr_reg_ibv_mr(...) {
    if (require_dmabuf) {
        goto dmabuf;
    }
    if (no_dmabuf_support) {
        goto no_dmabuf;
    }
dmabuf:
    // Try dmabuf fd/mr
update_dmabuf_support:
    // Depending on whether dmabuf succeeded, update the support status
no_dmabuf:
    // Fallback ibv_reg_mr if allowed. No need to update dmabuf support in this case.
}
```

Reply:
The problem with your pseudo code is that the different blocks doesn't have unique dependency relationship. For example,  for synapseai and FI_MR_DMABUF, I don't think we have to reach `update_dmabuf_support` and `no_dmabuf` blocks no matter whether dmabuf block succeeds or not. That ends up with you have conditional return between `dmabuf` and subsequent blocks which is more confusing than making explicit code sections for different hmem ifaces.

Reply:
Thinking about it again, it may not be that bad, I will review the latest change before making more comments here

Reply:
Refactored the logic after offline discussion. We will keep the current MR logic unchanged.